### PR TITLE
feat: add new rule to catch invalid character in multipart headers

### PR DESCRIPTION
**Pull request by airween**
Add a new rule `922130` which check if any MULTIPART header contains a non-ascii character.

Fixes 3MU-240701-1 by @luelueking for v4.

### CODE REVIEW

File: rules/REQUEST-922-MULTIPART-ATTACK.conf [hunk 1]
```diff
@@ -95,3 +95,26 @@ SecRule MULTIPART_PART_HEADERS "@rx content-transfer-encoding:(.*)" \
...
+    phase:2,\
+    block,\
+    capture,\
+    t:none,t:lowercase,\
+    msg:'Multipart header contains special characters which should not be used',\
+    logdata:'Matched Data: %{TX.0}',\
+    tag:'application-multi',\
+    tag:'language-multi',\
+    tag:'platform-multi',\
+    tag:'attack-deprecated-header',\
```
Comment (fzipi) [inline rules/REQUEST-922-MULTIPART-ATTACK.conf:113]:
```suggestion
    tag:'attack-multipart-header',\
```

**Review by fzipi — DISMISSED** [review]

File: tests/regression/tests/REQUEST-922-MULTIPART-ATTACK/922130.yaml [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    desc: "Positive test: special character in multipart header"
+    stages:
+      - input:
+          dest_addr: 127.0.0.1
+          port: 80
+          uri: "/post"
+          version: "HTTP/1.1"
+          encoded_request: "UE9TVCAvcG9zdCBIVFRQLzEuMQ0KQWNjZXB0OiAqLyoNCkNvbm5lY3Rpb246IGNsb3NlDQpDb250ZW50LUxlbmd0aDogMTg1DQpDb250ZW50LVR5cGU6IG11bHRpcGFydC9mb3JtLWRhdGE7IGJvdW5kYXJ5PWJvdW5kYXJ5DQpIb3N0OiBsb2NhbGhvc3QNClVzZXItQWdlbnQ6IE9XQVNQIENSUyB0ZXN0IGFnZW50DQoNCi0tYm91bmRhcnkNCg5Db250ZW50LURpc3Bvc2l0aW9uOiBmb3JtLWRhdGE7IG5hbWU9ImZpbGUiOyBmaWxlbmFtZT0iMS5waHAiDQpDb250ZW50LURpc3Bvc2l0aW9uOiBmb3JtLWRhdGE7IG5hbWU9InBvc3QiDQoNCjxwaHAgdmFyX2R1bXAoZmlsZV9nZXRfY29udGVudHMoIi9ldGMvcGFzc3dkIikpOw0KLS1ib3VuZGFyeS0t"
+        output:
+          log:
```
Comment (theseion) [inline tests/regression/tests/REQUEST-922-MULTIPART-ATTACK/922130.yaml:17]:
```suggestion
          log:
            # The rule will match when an older version of the engine doesn't fail parsing. In newer versions of the engine,
            # parsing will fail and the body will be considered invalid.
```

File: tests/regression/tests/REQUEST-922-MULTIPART-ATTACK/922130.yaml [hunk 3]
```diff
@@ -0,0 +1,41 @@
+---
+meta:
+  author: "Ervin Hegedus"
+  description: Test Multipart/form-data
+rule_id: 922130
+tests:
+  - test_id: 1
+    desc: "Positive test: special character in multipart header"
```
Comment (theseion) [inline tests/regression/tests/REQUEST-922-MULTIPART-ATTACK/922130.yaml:8]: Add the decoded contend (or at least a pseudo version of it) to the description, so its clear what the request looks like.
```suggestion
    desc: |
      Positive test: special character in multipart header.
      <decoded content>
```

**Review by theseion — CHANGES_REQUESTED** [review]

File: rules/REQUEST-922-MULTIPART-ATTACK.conf [hunk 4]
```diff
@@ -95,3 +95,26 @@ SecRule MULTIPART_PART_HEADERS "@rx content-transfer-encoding:(.*)" \
     ver:'OWASP_CRS/4.6.0-dev',\
     severity:'CRITICAL',\
     setvar:'tx.inbound_anomaly_score_pl1=+%{tx.critical_anomaly_score}'"
+
+# Multipart header names can't contain any characters outside of range 33 and 126.
+# RFC 2045 refers RFC 822 about the header syntax.
+# Note: this is in phase:2 because these are headers that come in the body
+SecRule MULTIPART_PART_HEADERS "@rx [^\x20-\x7E]" \
```
Comment (huberts90) [inline rules/REQUEST-922-MULTIPART-ATTACK.conf:102]: Let me ask if it is correct that the `x20` refers to the 32nd element in the ASCII table and the comment says that it should start from the 33rd.

File: tests/regression/tests/REQUEST-922-MULTIPART-ATTACK/922130.yaml (hunk 2 above)
Comment (airween) [reply tests/regression/tests/REQUEST-922-MULTIPART-ATTACK/922130.yaml:26]: Done.

File: tests/regression/tests/REQUEST-922-MULTIPART-ATTACK/922130.yaml (hunk 3 above)
Comment (airween) [reply tests/regression/tests/REQUEST-922-MULTIPART-ATTACK/922130.yaml]: Done

**Comment by airween** [comment]
I think I'm done with this PR.

File: tests/regression/tests/REQUEST-922-MULTIPART-ATTACK/922130.yaml [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+  description: Test Multipart/form-data
+rule_id: 922130
+tests:
+  - test_id: 1
+    desc: |
+      # Positive test: special character in multipart header
+      # Request is a multipart request with 2 Content-Disposition header,
+      # where the first one contains a 0x0E character
+      # --boundary
+      # Content-Disposition: form-data; name="file"; filename="1.php"
```
Comment (fzipi) [inline tests/regression/tests/REQUEST-922-MULTIPART-ATTACK/922130.yaml:13]:
```suggestion
      # \x0EContent-Disposition\x0E: form-data; name="file"; filename="1.php"
```

Comment (airween) [reply tests/regression/tests/REQUEST-922-MULTIPART-ATTACK/922130.yaml]: Done (see [f32c242](https://github.com/coreruleset/coreruleset/pull/3796/commits/f32c242813be6b5e9d6ea2dc8fad857949b47f87))

**Review by theseion — APPROVED** [review]

File: tests/regression/tests/REQUEST-922-MULTIPART-ATTACK/922130.yaml [hunk 6]
```diff
@@ -0,0 +1,52 @@
+---
+meta:
+  author: "Ervin Hegedus"
+  description: Test Multipart/form-data
+rule_id: 922130
+tests:
+  - test_id: 1
+    desc: |
+      # Positive test: special character in multipart header
```
Comment (theseion) [inline tests/regression/tests/REQUEST-922-MULTIPART-ATTACK/922130.yaml:9]: Remove the leading `#`.

**Review by theseion — CHANGES_REQUESTED** [review]

File: rules/REQUEST-922-MULTIPART-ATTACK.conf (hunk 4 above)
Comment (huberts90) [reply rules/REQUEST-922-MULTIPART-ATTACK.conf:102]: In my opinion, the relation between the comment and the rule is confusing. Either the comment should say:
> Multipart header names can't contain any characters outside of range 32 (space) and 126 (tilde).

or the rule should accept elements matching `[^\x21-\x7E]`.

Comment (theseion) [reply rules/REQUEST-922-MULTIPART-ATTACK.conf:102]: I agree. You're excluding the range `[0, 31]` and `[127,)`, but the comment says that you're excluding `[0, 32]` and `[127,)`.

```
# Multipart header names must only contain printable ASCII characters in the range of [33, 126], except colon.
# As `MULTIPART_PART_HEADERS` contains the full header line as a value, colon is still included in the match.
```

And the regex should be changed to what @huberts90 suggests. Nice find @huberts90!

**Review by theseion — CHANGES_REQUESTED** [review]

File: rules/REQUEST-922-MULTIPART-ATTACK.conf (hunk 4 above)
Comment (airween) [reply rules/REQUEST-922-MULTIPART-ATTACK.conf]: You're right, guys - I fixed it, see commit f377c6ac.

File: tests/regression/tests/REQUEST-922-MULTIPART-ATTACK/922130.yaml (hunk 6 above)
Comment (airween) [reply tests/regression/tests/REQUEST-922-MULTIPART-ATTACK/922130.yaml]: Fixed in 903974c8.

File: tests/regression/tests/REQUEST-922-MULTIPART-ATTACK/922130.yaml [hunk 7]
```diff
@@ -6,15 +6,14 @@ rule_id: 922130
...
-      #
-      # <?php var_dump(file_get_contents("/etc/passwd"));
-      # --boundary--
+      Positive test: special character in multipart header
+      Request is a multipart request with 2 Content-Disposition header,
+      where the first one contains a 0x0E character
+      --boundary
+      \x0EContent-Disposition: form-data; name="file"; filename="1.php"
+      Content-Disposition: form-data; name="post"
+            # <?php var_dump(file_get_contents("/etc/passwd"));
```
Comment (theseion) [inline tests/regression/tests/REQUEST-922-MULTIPART-ATTACK/922130.yaml:15]:
```suggestion
        <?php var_dump(file_get_contents("/etc/passwd"));
```

File: tests/regression/tests/REQUEST-922-MULTIPART-ATTACK/922130.yaml
Comment (theseion) [inline tests/regression/tests/REQUEST-922-MULTIPART-ATTACK/922130.yaml:1]: I think you should add four tests to check the boundaries, one for each of `\x20`, `\x21`,`\x126`, `\x127`.

**Review by theseion — CHANGES_REQUESTED** [review]

File: tests/regression/tests/REQUEST-922-MULTIPART-ATTACK/922130.yaml
Comment (airween) [reply tests/regression/tests/REQUEST-922-MULTIPART-ATTACK/922130.yaml:1]: Added in [2d89377](https://github.com/coreruleset/coreruleset/pull/3796/commits/2d89377d97cc918c9f6faf2b3624c2874b025506)

File: tests/regression/tests/REQUEST-922-MULTIPART-ATTACK/922130.yaml [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+          data: |
+            --boundary
+            Content-!Disposition: form-data; name="file"; filename="1.php"
+            Content-Disposition: form-data; name="post"
+
+            <?php var_dump(file_get_contents("/etc/passwd")); ?>
+            --boundary--
+        output:
+          log:
+            no_match_regex: 'Multipart parsing error: Multipart: Invalid part header \(contains invalid character\)|\[id "922130"\] \[msg "Multipart header contains special characters which should not be used"\]'
```
Comment (theseion) [inline tests/regression/tests/REQUEST-922-MULTIPART-ATTACK/922130.yaml:145]:
```suggestion
            no_match_regex: 'Multipart parsing error: Multipart: Invalid part header \(contains invalid character\)|\[id "922130"\] \[msg "Multipart header contains characters outside of valid range"\]'
```

File: tests/regression/tests/REQUEST-922-MULTIPART-ATTACK/922130.yaml [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+  - test_id: 7
+    desc: |
+      Positive test: special character in multipart header
```
Comment (theseion) [inline tests/regression/tests/REQUEST-922-MULTIPART-ATTACK/922130.yaml:148]:
```suggestion
      Positive test: special character in multipart header.
```

File: tests/regression/tests/REQUEST-922-MULTIPART-ATTACK/922130.yaml [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+      Request is a multipart request with 2 Content-Disposition header,
```
Comment (theseion) [inline tests/regression/tests/REQUEST-922-MULTIPART-ATTACK/922130.yaml:149]:
```suggestion
      Request is a multipart request with 2 Content-Disposition headers,
```

File: tests/regression/tests/REQUEST-922-MULTIPART-ATTACK/922130.yaml [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+      where the first one contains a 0x0E character
```
Comment (theseion) [inline tests/regression/tests/REQUEST-922-MULTIPART-ATTACK/922130.yaml:150]:
```suggestion
      where the first one contains a 0x7F character, the first invalid character after the valid range.
```

File: tests/regression/tests/REQUEST-922-MULTIPART-ATTACK/922130.yaml [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+      --boundary
+      Content~Disposition: form-data; name="file"; filename="1.php"
```
Comment (theseion) [inline tests/regression/tests/REQUEST-922-MULTIPART-ATTACK/922130.yaml:152]:
```suggestion
      Content\x7F-Disposition: form-data; name="file"; filename="1.php"
```

File: tests/regression/tests/REQUEST-922-MULTIPART-ATTACK/922130.yaml [hunk 13]
```diff
@@ -0,0 +1,169 @@
...
+          dest_addr: 127.0.0.1
+          port: 80
+          uri: "/post"
+          version: "HTTP/1.1"
+          encoded_request: "UE9TVCAvcG9zdCBIVFRQLzEuMQ0KQWNjZXB0OiAqLyoNCkNvbm5lY3Rpb246IGNsb3NlDQpDb250ZW50LUxlbmd0aDogMTg2DQpDb250ZW50LVR5cGU6IG11bHRpcGFydC9mb3JtLWRhdGE7IGJvdW5kYXJ5PWJvdW5kYXJ5DQpIb3N0OiBsb2NhbGhvc3QNClVzZXItQWdlbnQ6IE9XQVNQIENSUyB0ZXN0IGFnZW50DQoNCi0tYm91bmRhcnkNCkNvbnRlbnR/LURpc3Bvc2l0aW9uOiBmb3JtLWRhdGE7IG5hbWU9ImZpbGUiOyBmaWxlbmFtZT0iMS5waHAiDQpDb250ZW50LURpc3Bvc2l0aW9uOiBmb3JtLWRhdGE7IG5hbWU9InBvc3QiDQoNCjw/cGhwIHZhcl9kdW1wKGZpbGVfZ2V0X2NvbnRlbnRzKCIvZXRjL3Bhc3N3ZCIpKTsNCi0tYm91bmRhcnktLQ=="
+        output:
+          log:
+            # The rule will match when an older version of the engine doesn't fail parsing. In newer versions of the engine,
+            # parsing will fail and the body will be considered invalid.
+            match_regex: 'Multipart parsing error: Multipart: Invalid part header \(contains invalid character\)|\[id "922130"\] \[msg "Multipart header contains special characters which should not be used"\]'
```
Comment (theseion) [inline tests/regression/tests/REQUEST-922-MULTIPART-ATTACK/922130.yaml:168]:
```suggestion
            match_regex: 'Multipart parsing error: Multipart: Invalid part header \(contains invalid character\)|\[id "922130"\] \[msg "Multipart header contains characters outside of valid range"\]'
```

**Review by theseion — CHANGES_REQUESTED** [review]

File: tests/regression/tests/REQUEST-922-MULTIPART-ATTACK/922130.yaml [hunk 14]
```diff
@@ -72,12 +72,12 @@ tests:
...
-            no_match_regex: 'Multipart parsing error: Multipart: Invalid part header \(contains invalid character\)|\[id "922130"\] \[msg "Multipart header contains special characters which should not be used"\]'
+            no_match_regex: 'Multipart parsing error: Multipart: Invalid part header \(contains invalid character\)|\[id "922130"\] \[msg "Multipart header contains characters outside of valid range"\]'
   - test_id: 4
     desc: |
-      Positive test: special character in multipart header
-      Request is a multipart request with 2 Content-Disposition header,
-      where the first one contains a 0x0E character
+      Positive test: special character in multipart header.
+      Request is a multipart request with 2 Content-Disposition headers,
+      where the first one contains a 0x0E character, the last invalid character before the valid range.
```
Comment (theseion) [inline tests/regression/tests/REQUEST-922-MULTIPART-ATTACK/922130.yaml:80]:
```suggestion
      where the first one contains a 0x20 character, the last invalid character before the valid range.
```

File: tests/regression/tests/REQUEST-922-MULTIPART-ATTACK/922130.yaml [hunk 15]
```diff
@@ -95,9 +95,9 @@ tests:
           log:
             # The rule will match when an older version of the engine doesn't fail parsing. In newer versions of the engine,
             # parsing will fail and the body will be considered invalid.
-            match_regex: 'Multipart parsing error: Multipart: Invalid part header \(contains invalid character\)|\[id "922130"\] \[msg "Multipart header contains special characters which should not be used"\]'
+            match_regex: 'Multipart parsing error: Multipart: Invalid part header \(contains invalid character\)|\[id "922130"\] \[msg "Multipart header contains characters outside of valid range"\]'
   - test_id: 5
-    desc: "Negative test: no character from outside of range in multipart header; '!' is allowed (\x21)"
+    desc: "Negative test: no character from outside of range in multipart header; '!' is allowed (\x21, the first character of the range of valid characters))"
```
Comment (theseion) [inline tests/regression/tests/REQUEST-922-MULTIPART-ATTACK/922130.yaml:100]:
```suggestion
    desc: "Negative test: no character from outside of range in multipart header; '!' is allowed (\x21, the first character in the range of valid characters))"
```

**Review by theseion — CHANGES_REQUESTED** [review]

**Comment by fzipi** [comment]
JFYI: I'll wait until this one is merged with the definitive rule and test to do the backporting.

**Review by theseion — APPROVED** [review]

**Comment by airween** [comment]
@theseion - thanks for all of your help.
@fzipi, @huberts90 - do you want to take a review?

**Review by fzipi — APPROVED** [review]
LGTM. Thanks everyone 🙂
